draft-barnes-acme.md

```diff
@@ -516,18 +554,214 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
+
+In addition to the recovery mechanisms defined by ACME, individual client implementations may also offer implementation-specific recovery mechanisms.  For example, if a client creates account keys deterministically from a seed value, then this seed could be used to recover the account key by re-generating it.  Or an implementation could escrow an encrypted copy of the account key with a cloud storage provider, and give the encryption key to the user as a recovery value.
+
+### Token-Based Recovery
+
+With token-based recovery, the client proves to the server that it holds a secret value established in the initial registration transaction.  The client requests token-based recovery by sending a MAC over the new account key, using the recovery key from the initial registration.
+
+method (required, string):
+: The string "token"
+
+base (required, string):
+: The URI for the registration to be recovered.
+
+token (required, string):
+: A JSON-formatted JWS object using an HMAC algorithm, whose payload is the JWK representation of the public key of the new account key pair.
+
+~~~~~~~~~~
+
+POST /acme/recover-registration HTTP/1.1
+Host: example.com
+
+{
+  "resource": "recover-reg",
+  "method": "token",
+  "base": "https://example.com/acme/reg/asdf",
+  "token": {
+    "header": { "alg": "HS256" },
+    "payload": base64(JWK(newAccountKey)),
+    "signature": "..."
+  }
+}
+/* Signed as JWS, with new account key */
+
+~~~~~~~~~~
+
+On receiving such a request the server MUST verify that:
+
+* The base registration has a recovery key associated with it
+* The "alg" value represents a MAC algorithm
+* The MAC verifies using the recovery key
+* The JWK in the payload represents the new account key (i.e. the key used to verify the ACME message)
+
+If those conditions are met, and the recovery request is otherwise acceptable to the server, then the recovery process has succeeded.  The server creates a new registration resource based on the base registration and the new account key, and returns it on a 201 (Created) response, together with a Location header indicating a URI for the new registration.  If the recovery request is unsuccessful, the server returns an error response, such as 403 (Forbidden).
+
+~~~~~~~~~~
+
+HTTP/1.1 201 Created
+Content-Type: application/json
+Location: https://example.com/acme/reg/asdf
+Link: <https://example.com/acme/new-authz>;rel="next"
+Link: <https://example.com/acme/recover-reg>;rel="recover"
+Link: <https://example.com/acme/terms>;rel="terms-of-service"
+
+{
+  "key": { /* JWK from JWS header */ },
+
+  "contact": [
+    "mailto:cert-admin@example.com",
+    "tel:+12025551212"
```

Comment:
You mean the "contact" field?  That's part of the registration object, so since we're giving the registration to this new key, we give it the object.